File: 二分查找/1101 D天送货/Answer.cpp
class Solution {
public:
    int shipWithinDays(vector<int>& weights, int D) {
     int left=*max_element(weights.begin(),weights.end()),right=accumulate(weights.begin(),weights.end(),0);
     //定义二分查找的左右边界
     while(left<right)
     {
        int mid=(left+right)/2;
        int cur=0;//当前天运载量
        int need=1;//需要运送天数
         for(int load :weights)
         {
             if(cur+load>mid)
             {
                 need++;
                 cur=0;

             }
            cur+=load;//如果当前装载量大于标的值,则视为在新的一天进行装载,清空存货,天数+1;若未大于则进行实际装载
         
        }
        if(need>D){left=mid+1;}
         else{right=mid;}


    }
    return left;
    }
};

File: 二分查找/1101 D天送货/answer.py
class Solution:
    def shipWithinDays(self, weights: List[int], D: int) -> int:
        left=max(weights)
        right=sum(weights)
        while left<right :
            need=1
            cur=0
            mid=(left+right)//2
            for weight in weights:
                if cur+weight>mid:
                    cur=0
                    need=need+1

                cur+=weight

            if need>D :
                left=mid+1
            else:
                right=mid
        return left                
        
